group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java

```diff
@@ -0,0 +1,415 @@
+                return maybeTransitionFromRevokingToAssigningOrStable();
+
+            // Check if pending partitions have been freed up.
+            case ASSIGNING:
+                return maybeTransitionFromAssigningToAssigningOrStable();
+
+            // Nothing to do.
+            case STABLE:
+                return member;
+        }
+
+        return member;
+    }
+
+    /**
+     * Transitions to the initial state. Here we compute the Assigned,
+     * Revoking and Assigning sets.
+     *
+     * @return A new ConsumerGroupMember.
+     */
+    private ConsumerGroupMember transitionToInitialState() {
+        Map<Uuid, Set<Integer>> newAssignedSet = new HashMap<>();
+        Map<Uuid, Set<Integer>> newRevokingSet = new HashMap<>();
+        Map<Uuid, Set<Integer>> newAssigningSet = new HashMap<>();
+
+        // Compute the combined set of topics.
+        Set<Uuid> allTopicIds = new HashSet<>(targetAssignment.partitions().keySet());
+        allTopicIds.addAll(member.assignedPartitions().keySet());
+        allTopicIds.addAll(member.partitionsPendingRevocation().keySet());
+        allTopicIds.addAll(member.partitionsPendingAssignment().keySet());
+
+        for (Uuid topicId : allTopicIds) {
+            Set<Integer> target = targetAssignment.partitions()
+                .getOrDefault(topicId, Collections.emptySet());
+            Set<Integer> currentAssignedPartitions = member.assignedPartitions()
+                .getOrDefault(topicId, Collections.emptySet());
+            Set<Integer> currentRevokingPartitions = member.partitionsPendingRevocation()
+                .getOrDefault(topicId, Collections.emptySet());
+
+            // Assigned_1 = (Assigned_0 + Revoking_0) /\ Target
+            Set<Integer> newAssignedPartitions = new HashSet<>(currentAssignedPartitions);
+            newAssignedPartitions.addAll(currentRevokingPartitions);
+            newAssignedPartitions.retainAll(target);
+
+            // Revoking_1 = (Assigned_0 + Revoking_0) - Assigned_1
+            Set<Integer> newRevokingPartitions = new HashSet<>(currentAssignedPartitions);
+            newRevokingPartitions.addAll(currentRevokingPartitions);
+            newRevokingPartitions.removeAll(newAssignedPartitions);
+
+            // Assigning_1 = Target - Assigned_1
+            Set<Integer> newAssigningPartitions = new HashSet<>(target);
+            newAssigningPartitions.removeAll(newAssignedPartitions);
+
+            if (!newAssignedPartitions.isEmpty()) {
+                newAssignedSet.put(topicId, newAssignedPartitions);
+            }
+
+            if (!newRevokingPartitions.isEmpty()) {
+                newRevokingSet.put(topicId, newRevokingPartitions);
+            }
+
+            if (!newAssigningPartitions.isEmpty()) {
+                newAssigningSet.put(topicId, newAssigningPartitions);
+            }
+        }
+
+        if (!newRevokingSet.isEmpty()) {
+            // If the revoking set is not empty, we transition to Revoking and we
```

Comment:
Is this transition done in the response or somewhere else?

Reply:
It is done by the code right after. The member automatically transitions to Revoking when `newPartitionsPendingRevocation` is not empty.

Reply:
Updated the comments. However, I feel like the link to MemberState is not really needed as we already explain all the states in the main javadoc.